sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,734 @@
+	sof_pdata->machine = mach;
+	sof_pdata->desc = &virt_desc;
+	sof_pdata->dev = dev;
+	sof_pdata->vfe = vdev->priv;
+	sof_pdata->tplg_filename_prefix = virt_desc.default_tplg_path;
+
+	ret = snd_sof_device_probe(dev, sof_pdata);
+	if (ret < 0) {
+		dev_err(dev, "Cannot register device sof-audio. Error %d\n",
+			ret);
+	} else {
+		sof_virtio_vfe_init(dev_get_drvdata(dev), vdev->priv);
+
+		dev_dbg(dev, "created machine %s\n",
+			dev_name(&sof_pdata->pdev_mach->dev));
+	}
+
+	/* allow runtime_pm */
+	pm_runtime_set_autosuspend_delay(dev, SND_SOF_SUSPEND_DELAY_MS);
+	pm_runtime_use_autosuspend(dev);
+	pm_runtime_enable(dev);
+
+	return ret;
+}
+
+/* Probe and remove. */
+static int sof_vfe_probe(struct virtio_device *vdev)
+{
+	struct virtqueue *vqs[SOF_VIRTIO_NUM_OF_VQS];
+	/* the processing callback number must be the same as the vqueues.*/
+	vq_callback_t *cbs[SOF_VIRTIO_NUM_OF_VQS] = {
+		sof_vfe_cmd_tx_done,
+		sof_vfe_psn_handle_rx,
+		sof_vfe_handle_data,
+	};
+	struct device *dev = &vdev->dev;
+	struct sof_vfe *vfe;
+	int ret;
+
+	/*
+	 * The below two shouldn't be necessary, it's done in
+	 * virtio_pci_modern_probe() by calling dma_set_mask_and_coherent()
+	 */
+
+	ret = dma_set_coherent_mask(dev, DMA_BIT_MASK(64));
+	if (ret < 0)
+		ret = dma_set_coherent_mask(dev, DMA_BIT_MASK(32));
+	if (ret < 0)
+		dev_warn(dev, "failed to set DMA mask: %d\n", ret);
+
+	vfe = devm_kzalloc(dev, sizeof(*vfe), GFP_KERNEL);
+	if (!vfe)
+		return -ENOMEM;
+
+	vdev->priv = vfe;
+
+	INIT_WORK(&vfe->posn_update_work, sof_vfe_posn_update);
+	init_completion(&vfe->completion);
+
+	/* create virt queue for vfe to send/receive IPC message. */
+	ret = virtio_find_vqs(vdev, SOF_VIRTIO_NUM_OF_VQS,
+			      vqs, cbs, sof_vq_names, NULL);
+	if (ret) {
+		dev_err(dev, "error: find vqs fail with %d\n", ret);
+		return ret;
+	}
+
+	/* virtques */
+	vfe->ipc_cmd_vq = vqs[SOF_VIRTIO_IPC_CMD_VQ];
+	vfe->ipc_psn_vq = vqs[SOF_VIRTIO_IPC_PSN_VQ];
+	vfe->data_vq = vqs[SOF_VIRTIO_DATA_VQ];
+
+	virtio_device_ready(vdev);
+
+	return sof_vfe_init(vdev);
+}
+
+static void sof_vfe_remove(struct virtio_device *vdev)
+{
+	/* free virtio resurces and unregister device */
+	struct sof_vfe *vfe = vdev->priv;
+
+	vdev->config->reset(vdev);
+	vdev->config->del_vqs(vdev);
+	cancel_work_sync(&vfe->posn_update_work);
+
+	/* unregister the SOF device */
+	snd_sof_device_remove(&vdev->dev);
+
+	return;
+}
+
+static void virtaudio_config_changed(struct virtio_device *vdev)
+{
+}
+
+/*
+ * Need to patch QEMU to create a virtio audio device, e.g. per
+ * -device virtio-snd-pci,snd=snd0 where Device ID must be
+ * 0x1040 + VIRTIO_ID_DSP and Vendor ID = PCI_VENDOR_ID_REDHAT_QUMRANET
+ */
+static const struct virtio_device_id id_table[] = {
+	{VIRTIO_ID_DSP, VIRTIO_DEV_ANY_ID},
+	{0},
+};
+
+/*
+ * TODO: There still need a shutdown to handle the case the UOS
+ * is poweroff, restart.
+ */
+
+static int sof_vfe_runtime_suspend(struct device *dev)
+{
+	return 0;
```

Comment:
@lyakh I am equating the guest to a "client" when looking at it from the perspective of what I am working on. Im this case, if you do restore pipelines in resume you have to also take them down during suspend, no? Unless you are assuming that the DSP is also suspended and in D3 when this virtual device suspends. Could you please help me understand?

Reply:
@ranj063 RT-PM hasn't been tested, it comes from the original ACRN implementation, we probably should remove it until we test and fix it.

Reply:
@ranj063 now that runtime-PM has been tested to work too, yes, there are two possibilities: (1) the DSP is also suspended because while SOF on the guest VM entered runtime suspend, the same happened on the host. (2) the DSP isn't suspended on the host. For this I now added a new IPC to verify whether the host has been through a suspend / resume cycle since the last time the guest used it, and if not - the guest doesn't send the pipeline data to the host.